File: python/docs/examples/browser-use/browser/browser_command_args.py
"""
Example demonstrating Browser Launch with Custom Command Arguments and
go to Default Navigation URL with AgentBay SDK.

This example shows how to configure browser with custom command arguments
and go to default navigation URL:
- Create AIBrowser session with custom command arguments and go to default navigation URL
- Use playwright to connect to AIBrowser instance through CDP protocol
- Verify the browser navigated to the default URL
- Test custom command arguments effects
"""

import os
import asyncio

from agentbay import AgentBay
from agentbay.session_params import CreateSessionParams
from agentbay.browser.browser import BrowserOption, BrowserFingerprint
from playwright.async_api import async_playwright


async def main():
    # Get API key from environment variable
    api_key = os.getenv("AGENTBAY_API_KEY")
    if not api_key:
        print("Error: AGENTBAY_API_KEY environment variable not set")
        return

    # Initialize AgentBay client
    print("Initializing AgentBay client...")
    agent_bay = AgentBay(api_key=api_key)

    # Create a session
    print("Creating a new session...")
    params = CreateSessionParams(
        image_id="linux_latest"
    )
    session_result = agent_bay.create(params)

    if session_result.success:
        session = session_result.session
        print(f"Session created with ID: {session.session_id}")

        # Create browser option with user defined cmd args and default navigate url
        browser_option = BrowserOption(
            cmd_args=["--disable-features=PrivacySandboxSettings4"],
            default_navigate_url="chrome://version/",
        )

        print("Browser configuration:")
        print("- Command arguments:", browser_option.cmd_args)
        print("- Default navigate URL:", browser_option.default_navigate_url)

        if await session.browser.initialize_async(browser_option):
            print("Browser initialized successfully")
            
            # Get browser endpoint URL
            endpoint_url = session.browser.get_endpoint_url()
            print(f"endpoint_url = {endpoint_url}")

            # Use Playwright to connect and validate
            async with async_playwright() as p:
                browser = await p.chromium.connect_over_cdp(endpoint_url)
                context = browser.contexts[0]
                page = context.pages[0]

                try:
                    # Check if browser navigated to default URL
                    print("\n--- Check Default Navigation ---")
                    await asyncio.sleep(2)  # Wait for navigation
                    current_url = page.url
                    print(f"Current URL: {current_url}")
                    
                    if "chrome://version/" in current_url:
                        print("✓ Browser successfully navigated to default URL")
                    else:
                        print("✗ Browser did not navigate to default URL")

                    # Test command arguments effect by checking Chrome version page
                    if "chrome://version/" in current_url:
                        print("\n--- Check Chrome Version Info ---")
                        version_info = await page.evaluate("""
                            () => {
                                const versionElement = document.querySelector('#version');
                                const commandLineElement = document.querySelector('#command_line');
                                return {
                                    version: versionElement ? versionElement.textContent : 'Not found',
                                    commandLine: commandLineElement ? commandLineElement.textContent : 'Not found'
                                };
                            }
                        """)
                        
                        print(f"Chrome Version: {version_info['version']}")
                        print(f"Command Line: {version_info['commandLine']}")
                        
                        if "--disable-features=PrivacySandboxSettings4" in version_info['commandLine']:
                            print("✓ Custom command argument found in browser")
                        else:
                            print("✗ Custom command argument not found in browser")

                    await asyncio.sleep(3)
                finally:
                    await browser.close()
                    session.browser.destroy()
        else:
            print("Failed to initialize browser")

        # Clean up session
        agent_bay.delete(session)
    else:
        print("Failed to create session", session_result.error_message)
    
if __name__ == "__main__":
    asyncio.run(main())

File: typescript/docs/examples/browser-use/browser/browser-command-args.ts
/**
 * Example demonstrating Browser Launch with Custom Command Arguments and 
 * go to Default Navigation URL with AgentBay SDK.
 *
 * This example shows how to configure browser with custom command arguments
 * and go to default navigation URL:
 * - Create AIBrowser session with custom command arguments and go to default navigation URL
 * - Use playwright to connect to AIBrowser instance through CDP protocol
 * - Verify the browser navigated to the default URL
 * - Test custom command arguments effects
 */

// @ts-nocheck
import { AgentBay, CreateSessionParams } from 'wuying-agentbay-sdk';
import { BrowserOption } from 'wuying-agentbay-sdk';
import { chromium } from 'playwright';

async function main(): Promise<void> {
    // Get API key from environment variable
    const apiKey = process.env.AGENTBAY_API_KEY;
    if (!apiKey) {
        console.log('Error: AGENTBAY_API_KEY environment variable not set');
        return;
    }

    try {
        // Initialize AgentBay client
        console.log('Initializing AgentBay client...');
        const agentBay = new AgentBay({ apiKey });

        // Create a session
        console.log('Creating a new session...');
        const params: CreateSessionParams = {
            imageId: 'linux_latest',
        };
        const sessionResult = await agentBay.create(params);

        if (!sessionResult.success) {
            console.log('Failed to create session');
            return;
        }

        const session = sessionResult.session;
        console.log(`Session created with ID: ${session.sessionId}`);

        // Create browser option with user defined cmd args and default navigate url
        const browserOption: BrowserOption = {
            cmdArgs: ['--disable-features=PrivacySandboxSettings4'],
            defaultNavigateUrl: 'chrome://version/',
        };

        console.log('Browser configuration:');
        console.log('- Command arguments:', browserOption.cmdArgs);
        console.log('- Default navigate URL:', browserOption.defaultNavigateUrl);

        const initialized = await session.browser.initializeAsync(browserOption);
        if (initialized) {
            const endpointUrl = await session.browser.getEndpointUrl();
            console.log('endpoint_url =', endpointUrl);

            const browser = await chromium.connectOverCDP(endpointUrl);
            const context = browser.contexts()[0];
            console.log('page count =', context.pages().length);
            const page = context.pages()[0];

            try {
                // Check if browser navigated to default URL
                console.log('\n--- Check Default Navigation ---');
                await page.waitForTimeout(2000); // Wait for navigation
                const currentUrl = page.url();
                console.log('Current URL:', currentUrl);
                
                if (currentUrl.includes('chrome://version/')) {
                    console.log('✓ Browser successfully navigated to default URL');
                } else {
                    console.log('✗ Browser did not navigate to default URL');
                }

                // Test command arguments effect by checking Chrome version page
                if (currentUrl.includes('chrome://version/')) {
                    console.log('\n--- Check Chrome Version Info ---');
                    const versionInfo = await page.evaluate(() => {
                        const versionElement = document.querySelector('#version');
                        const commandLineElement = document.querySelector('#command_line');
                        return {
                            version: versionElement ? versionElement.textContent : 'Not found',
                            commandLine: commandLineElement ? commandLineElement.textContent : 'Not found'
                        };
                    });
                    
                    console.log('Chrome Version:', versionInfo.version);
                    console.log('Command Line:', versionInfo.commandLine);
                    
                    if (versionInfo.commandLine.includes('--disable-features=PrivacySandboxSettings4')) {
                        console.log('✓ Custom command argument found in browser');
                    } else {
                        console.log('✗ Custom command argument not found in browser');
                    }
                }

                await page.waitForTimeout(3000);
            } finally {
                await browser.close();
            }
            await session.browser.destroy();
        }

        // Clean up session
        await agentBay.delete(session);
    } catch (error) {
        console.error('Error:', error);
    }
}

if (require.main === module) {
    main().catch(console.error);
}
